Realm/RLMDictionary.mm

```diff
@@ -0,0 +1,541 @@
+        default:
+            return false;
+    }
+}
+
+void RLMDictionaryValidateMatchingObjectType(__unsafe_unretained RLMDictionary *const dictionary,
+                                             __unsafe_unretained id const key,
+                                             __unsafe_unretained id const value) {
+    if (!key) {
+        @throw RLMException(@"Invalid nil key for dictionary expecting key of type '%@'.",
+                            dictionary->_objectClassName ?: RLMTypeToString(dictionary.keyType));
+    }
+    if (!RLMValidateValue(key, dictionary.keyType, false, false, nil)) {
+        @throw RLMException(@"Invalid key '%@' of type '%@' for expected type '%@'.",
+                            key, [key class], RLMTypeToString(dictionary.keyType));
+    }
+    if (!value) {
+        return;
+    }
+    if (dictionary->_type != RLMPropertyTypeObject) {
+        if (!RLMValidateValue(value, dictionary->_type, dictionary->_optional, false, nil)) {
+            @throw RLMException(@"Invalid value '%@' of type '%@' for expected type '%@%s'.",
+                                value, [value class], RLMTypeToString(dictionary->_type),
+                                dictionary->_optional ? "?" : "");
+        }
+        return;
+    }
+    auto valueObject = RLMDynamicCast<RLMObjectBase>(value);
+    if (!valueObject) {
+        @throw RLMException(@"Value of type '%@' does not match RLMDictionary value type '%@'.",
+                            [value className], dictionary->_objectClassName);
+    }
+    if (!valueObject->_objectSchema) {
+        @throw RLMException(@"Object cannot be inserted unless the schema is initialized. "
+                            "This can happen if you try to insert objects into a RLMDictionary / Map from a default value or from an overriden unmanaged initializer (`init()`) or if the key is uninitialized.");
+    }
+    if (![dictionary->_objectClassName isEqualToString:valueObject->_objectSchema.className]) {
+        @throw RLMException(@"Value of type '%@' does not match RLMDictionary value type '%@'.",
+                            valueObject->_objectSchema.className, dictionary->_objectClassName);
+    }
+}
+
+static void changeDictionary(__unsafe_unretained RLMDictionary *const dictionary,
+                             dispatch_block_t f) {
+    if (!dictionary->_backingCollection) {
+        dictionary->_backingCollection = [NSMutableDictionary new];
+    }
+    if (RLMObjectBase *parent = dictionary->_parentObject) {
+        [parent willChangeValueForKey:dictionary->_key];
+        f();
+        [parent didChangeValueForKey:dictionary->_key];
+    }
+    else {
+        f();
+    }
+}
+
+#pragma mark - Unmanaged RLMDictionary implementation
+
+- (RLMRealm *)realm {
+    return nil;
+}
+
+- (NSUInteger)count {
+    return _backingCollection.count;
+}
+
+- (NSArray *)allKeys {
+    return _backingCollection.allKeys;
+}
+
+- (NSArray *)allValues {
+    return _backingCollection.allValues;
+}
```

Comment:
These need a `?: @[]` at the end so that we return an empty array rather than nil if _backingCollection isn't initialized yet.